### PR TITLE
Collider optimization

### PR DESCRIPTION
**Pull request by pca006132**
Unrelated:

1. Fixed tracy (again the cmake interface vs public issue)
2. Moved some APIs from `Vec` to `VecView`.

Optimizations:

1. Only do collision check once, use thread local buffers to store the result and merge in parallel when everything is completed. Note that this requires making the vector copy sequential, so I added some flags to help.
2. Allow reusing collider result buffer, reduce reallocation during complex polygon triangulation.
3. Some low level optimization.

This results in about 8% time reduction for `perfTest` (2.5s to 2.3s), and 10% time reduction for running triangulation tests (~1900ms to ~1700ms)

### CODE REVIEW

File: src/polygon.cpp [hunk 1]
```diff
@@ -489,7 +489,8 @@ class EarClip {
     // values < -precision so they will never affect validity. The first
     // totalCost is designed to give priority to sharper angles. Any cost < (-1
     // - precision) has satisfied the Delaunay condition.
-    double EarCost(double precision, const IdxCollider &collider) const {
+    double EarCost(double precision, const IdxCollider &collider,
+                   SparseIndices &toTest) const {
```
Comment (pca006132) [inline src/polygon.cpp:493]: Thinking about it, I should probably bundle `toTest` to `IdxCollider`.

File: src/polygon.cpp [hunk 2]
```diff
@@ -502,27 +504,32 @@ class EarClip {
...
-          totalCost = std::max(totalCost, cost);
+          return cost;
         }
-      }
-
+        return std::numeric_limits<double>::lowest();
+      };
+      totalCost = transform_reduce(
+          countAt(0), countAt(collider.ind.size()), totalCost,
+          [](double a, double b) { return std::max(a, b); }, f);
```
Comment (elalish) [inline src/polygon.cpp:531]: If `f` is only being used in one place, should we just put it here directly?

File: src/vec.h [hunk 3]
```diff
@@ -145,25 +143,31 @@ class Vec : public VecView<T> {
...
       this->ptr_[this->size_++] = val_copy;
       return;
     }
     this->ptr_[this->size_++] = val;
   }
 
-  void reserve(size_t n) {
+  inline void extend(size_t n, bool seq = false) {
+    if (this->size_ + n >= capacity_)
+      reserve(capacity_ == 0 ? 128 : capacity_ * 2, seq);
```
Comment (elalish) [inline src/vec.h:159]: Is this right? If `n` is much larger than `capacity_`, will it actually increase enough?

**Review by elalish — CHANGES_REQUESTED** [review]
Thanks, looks great!

File: src/vec.h (hunk 3 above)
Comment (pca006132) [reply src/vec.h:159]: Nice catch, I copied the code from `push_back` and forgot that n can be large.

**Review by elalish — APPROVED** [review]
LGTM, thanks! Any idea what's wrong with the Black formatter?

**Comment by pca006132** [comment]
Hmmm no idea, it seems that everything somehow went wrong. We can try to run the CI again...

**Comment by pca006132** [comment]
The weird thing is that it starts to fail after the last commit, but the last commit should not affect anything...

**Comment by pca006132** [comment]
I will investigate this later this week.

**Comment by pca006132** [comment]
Just got some time, I still don't understand why the previous test failure can happen, but I guess it is caused by the cmake changes in the master branch. The weird thing is that I haven't merged the changes at [f632177](https://github.com/elalish/manifold/pull/978/commits/f6321777acb4d0c082907256b8fb16696d616d5f). Anyway, the tests pass now.

For the black formatter error, I think this is caused by the latest version (released a day ago) no longer supports python 3.8, and Ubuntu 20.04 ship with python 3.8 by default. Should be fixed by installing python 3.12. (which I did previously for the cmake formatter because I forgot Ubuntu ships with python preinstalled...)

**Review by elalish — APPROVED** [review]
Thanks!

**Pull request merged**